### PR TITLE
Add ability to import Collections with metadata

### PR DESCRIPTION
**Pull request by bkiahstroud**
# Summary

Add ability to import collections with metadata when `:model` or model mappings are set to Collection **(CSV only)** 

| model | source_identifier | title | description |
| --- | --- | --- | --- |
| Collection | collection_1 | First Collection | Lorem ipsum |

Retains ability to import Collections by title when importing works 

| model | source_identifier | title | collection |
| --- | --- | --- | --- |
| Work | work_1 | First Work | Parent Collection |

### CODE REVIEW

File: app/parsers/bulkrax/csv_parser.rb [hunk 1]
```diff
@@ -8,14 +8,26 @@ def self.export_supported?
...
+      importer.parser_fields['total'] = csv_data.count
+      importer.save
+      @records ||= csv_data.map { |record_data| entry_class.data_for_entry(record_data, nil) }
+    end
+
     def collections
-      # does the CSV contain a collection column?
-      return [] unless (import_fields & [:collection, :collections]).any?
+      col_mapping = Bulkrax.collection_field_mapping[self.entry_class.to_s]&.to_sym || :collection
+      model_mappings = Bulkrax.field_mappings[self.class.to_s]&.dig('model', :from) || []
```
Comment (alishaevn) [inline app/parsers/bulkrax/csv_parser.rb:22]: the `self.class.to_s` call in csv_entry.rb evaluations to "Class". will this not do the same?

File: app/parsers/bulkrax/oai_dc_parser.rb [hunk 2]
```diff
@@ -119,6 +119,11 @@ def collections_total
       end
     end
 
+    # TODO: change to differentiate between collection and work records when adding ability to import collection metadata
```
Comment (alishaevn) [inline app/parsers/bulkrax/oai_dc_parser.rb:122]: isn't this TODO outdated?

File: app/parsers/bulkrax/csv_parser.rb (hunk 1 above)
Comment (bkiahstroud) [reply app/parsers/bulkrax/csv_parser.rb:22]: It shouldn't. This is an instance method, so `self` in this case is an instance of `CsvParser`: 
```ruby
self
=> #<Bulkrax::CsvParser:0x00005602f005e2b0 ...
self.class
=> Bulkrax::CsvParser
```

I think the method in the `CsvEntry` as a class method, so `self` in that case would be the class: 
```ruby
self
=> Bulkrax::CsvEntry
self.class
=> Class
```

**Review by alishaevn — APPROVED** [review]

File: app/factories/bulkrax/object_factory.rb [hunk 3]
```diff
@@ -213,6 +212,10 @@ def transform_attributes
       @transform_attributes
     end
 
+    def parser
+      Entry.find_by(identifier: source_identifier_value)&.parser || ApplicationParser
```
Comment (sephirothkod) [inline app/factories/bulkrax/object_factory.rb:216]: can we lazy load this so we aren't making multiple calls?

File: app/models/bulkrax/csv_collection_entry.rb [hunk 4]
```diff
@@ -7,13 +7,27 @@ def factory_class
     end
 
     def build_metadata
-      self.parsed_metadata = self.raw_metadata
+      self.parsed_metadata = {}
+      self.parsed_metadata[work_identifier] = self.identifier
```
Comment (orangewolf) [inline app/models/bulkrax/csv_collection_entry.rb:11]: where does identifier come from?

Comment (bkiahstroud) [reply app/models/bulkrax/csv_collection_entry.rb:12]: [CsvParser#unique_collection_identifier](https://github.com/samvera-labs/bulkrax/pull/364/files#diff-ccbdc46eb1fafd2a6aec08c257c2589531c46fee5d793832a85bf0653f754490R275-R284). I added a comment to the top of the new method that handles this explaining why we're using `self.identifier`

**Review by orangewolf — APPROVED** [review]

**Pull request merged**